docs/development-guide/extending-gaffer.md

```diff
@@ -0,0 +1,217 @@
+which implements the appropriate `Handler` interface (in this case the
+`OutputOperationHandler`). It takes in the `ToSet` Operation that has been built
+with all the options etc. defined and then does the actual `doOperation()` logic to
+stream our Iterable's contents into a `LinkedHashSet`. This and most other
+handlers live inside the [Store library](./project-structure/components/store.md).
+
+#### How does Gaffer know which Handler to use?
+
+Handlers are registered in the `Store` class via the `addOpHandlers()` method,
+specifically for our example  `addOperationHandler(ToSet.class, new
+ToSetHandler<>())` is called to register it. Once this is done we can then
+use our `Operation` and its `Handler` in a user defined `OperationChain` to effect
+some manipulation/query on our graph data.
+
+You can alternatively use the an operations declarations config file (typically called
+`operationsDeclarations.json`) which can be used to define the Operation and its Handler
+like below:
+
+```json
+{
+    "operations": [
+        {
+            "operation": "uk.gov.gchq.gaffer.operation.impl.output.ToSet",
+            "handler": {
+                "class": "uk.gov.gchq.gaffer.store.operation.handler.output.ToSetHandler"
+            }
+        }
+    ]
+}
+```
+
+### Writing an Element/Object Generator
+
+There are two types of generators that define custom binding for taking a domain
+object and transforming it into a Gaffer `Element` for the `ElementGenerator` and
+likewise the reverse with an `ObjectGenerator`.
+
+Living within the Core/Data section of the repository these generator interfaces
+are quite simple and as with `Operation` are extended by other interfaces to
+provide a more nuanced set of interfaces you could choose to use over the core
+Element/Object generator interfaces.
+
+As an example we will walk through the one of the existing generators that can
+take a Neo4j CSV file and convert the entries to Gaffer `Elements`.
+
+```java
+/**
+ * An {@link OpenCypherCsvElementGenerator}s that will generate
+ * Gaffer Elements from Neo4j CSV strings.
+ */
+@Since("2.0.0")
+@Summary("Generates elements from a Neo4j CSV")
+public class Neo4jCsvElementGenerator extends OpenCypherCsvElementGenerator {
+    @Override
+    protected LinkedHashMap<String, String> getFields() {
+        return new Neo4jCsvGenerator().getFields();
+    }
+}
+```
+
+This class may appear very simple but the real functionality for this generator
+is in the `OpenCypherCsvElementGenerator` that it extends. This generator is an
+abstract class that implements the `ElementGenerator` interface and has a lot of
+functionality for handling Open Cypher CSVs.
+
+You can view the [full class in the repository](https://github.com/gchq/Gaffer/blob/develop/core/data/src/main/java/uk/gov/gchq/gaffer/data/generator/OpenCypherCsvElementGenerator.java) which has the logic for converting the CSV format
+but the key bit to take note of is the `apply(final Iterable<? extends String>
```

Comment:
Line is much longer than the others. Can probably shuffle the text after the link down to the next line.